### PR TITLE
Jocelyn july 15-17

### PR DESCRIPTION
**Pull request by jociet101**
# July 15-17 PR

# 👩🏻‍🏫 What does this PR do?

### Refactoring
- [x] Improved readability such as moving large code chunks to their own methods and adding ternary operators when suitable as per Alberto's feedback
- [x] Created Helpers file to move helper methods from Constants file

### API Manager
- [x] implement Reverse Geocoding method to query from Geoapify API

### Select Map view
- [x] Implement feature where user can long press on map to choose location
- [x] Implement button and action so user can use current location as origin location for filtering or a session

### Map pins
- [x] Implement MKAnnotationViews for all sessions, pins appear on map and user can select to view session details

### Session details
- [x] Add automatic refresh of UI when user joins or leaves session

### Other
- [x] Fetch sports list from Decathalon Sports API instead of hard coding
- [x] Fixed some warnings as per Carmen's feedback

### CODE REVIEW

File: Playmate/Utilities/Constants.m [hunk 1]
```diff
@@ -222,46 +259,32 @@ + (NSArray *)sportsList:(BOOL)needAll {
...
     NSMutableArray *sports = [[NSMutableArray alloc] init];
     
-    // TODO: Need to debug this for getting all sports later
-    /*
+    if (needAll) [sports addObject:@"All"];
+    
     APIManager *manager = [APIManager new];
     [manager getSportsListWithCompletion:^(NSDictionary *list, NSError *error) {
         if (error != nil) {
             NSLog(@"%@", error.localizedDescription);
```
Comment (albertomota-codes) [inline Playmate/Utilities/Constants.m:267]: should we alert the user there was an issue in the backend?

File: Playmate/Utilities/Constants.m [hunk 2]
```diff
@@ -222,46 +259,32 @@ + (NSArray *)sportsList:(BOOL)needAll {
...
-    */
-    
-    if (needAll) {
-        [sports addObject:@"All"];
-    }
     
-    [sports addObject:@"Tennis"];
-    [sports addObject:@"Basketball"];
-    [sports addObject:@"Golf"];
     return (NSArray *)sports;
```
Comment (albertomota-codes) [inline Playmate/Utilities/Constants.m:279]: I like more to do this: 
`[NSArray arrayWithArray:mutableArray]`

This way you really return a NSArray, instead of a NSMutableArray that is just casted as NSArray. 

It is super minor, but a little bit more correct.

File: Playmate/Utilities/Constants.m [hunk 3]
```diff
@@ -334,16 +339,30 @@ + (NSString *)emptyListPlaceholderTitle {
...
-+ (NSString *)emptyRequestsPlaceholderTitle {
++ (NSString *)emptyIncomingRequestsPlaceholderTitle {
     return @"No Incoming Friend Requests";
 }
 
 + (NSString *)emptyOutgoingRequestsPlaceholderTitle {
     return @"No Outgoing Friend Requests";
 }
 
++ (NSDictionary *)descriptionAttributes {
```
Comment (albertomota-codes) [inline Playmate/Utilities/Constants.m:350]: 💙

File: Playmate/Utilities/Helpers.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import "Session.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface Helpers : NSObject
+
++ (UIImage *)resizeImage:(UIImage *)image withDimension:(int)dimension;
```
Comment (albertomota-codes) [inline Playmate/Utilities/Helpers.h:16]: 💙

File: Playmate/View Controllers/Create Tab/SelectMapViewController.m [hunk 5]
```diff
@@ -68,35 +96,25 @@ - (void)handleAlert:(NSError * _Nullable)error withTitle:(NSString *)title andOk
...
-                MKCoordinateRegion region = MKCoordinateRegionMake(centerCoord, MKCoordinateSpanMake(0.1, 0.1));
-                [self.mapView setRegion:region animated:false];
-                
-                MKPointAnnotation *annotation = [[MKPointAnnotation alloc] init];
-                [annotation setCoordinate:centerCoord];
-                [annotation setTitle:loc.locationName];
-                [self.mapView addAnnotation:annotation];
-                
-                // send location back to filters or create view controller
+                [self dropPinOnMapAt:centerCoord withName:loc.locationName];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Create Tab/SelectMapViewController.m:117]: I like how you create functions to avoid huge complex logic inside the closure. 

Should we worry about  calling `self` inside a closure? due to memory retain cycle? 
Did you got any warning in Xcode?

File: Playmate/View Controllers/Home Tab/CalendarViewController.m [hunk 6]
```diff
@@ -154,16 +154,15 @@ - (NSInteger)calendar:(FSCalendar *)calendar numberOfEventsForDate:(NSDate *)dat
...
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
      
      if ([sender isMemberOfClass:[SessionCell class]]) {
          
          NSIndexPath *indexPath = [self.tableView indexPathForCell:sender];
          
          Session* data = self.sessionList[indexPath.section];
-         SessionDetailsViewController *VC = [segue destinationViewController];
-         VC.sessionDetails = data;
+         SessionDetailsViewController *vc = [segue destinationViewController];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/CalendarViewController.m:164]: 💙

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 7]
```diff
@@ -14,7 +14,7 @@
 #import "CalendarViewController.h"
 #import "ProfileViewController.h"
 
-@interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZNEmptyDataSetSource, DZNEmptyDataSetDelegate>
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:17]: How come the table view is still working? 

Did you had the protocols also in the .h?

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 8]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
+	PFQuery *query = [PFQuery queryWithClassName:@"SportsSession"];
+	query.limit = 20;
+
+	[query orderByAscending:@"occursAt"];
+
+	// fetch data asynchronously
+	[query findObjectsInBackgroundWithBlock:^(NSArray *sessions, NSError *error) {
+	         if (sessions != nil) {
+
+			 [self.sessionList removeAllObjects];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:72]: why did the indentation changed?

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 9]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
-                break;
-            }
-        }
-    }
-}
+	NSMutableArray *tempList = (NSMutableArray *)sessions;
+	PFUser *currUser = [[PFUser currentUser] fetchIfNeeded];
 
-#pragma mark - Empty table view protocol methods
+	for (Session *sesh in tempList) {
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:88]: Try to avoid abbreviations. 

I know they are tempting while writting, but after a couple of months when you come back to the code, you will  be glad to have used the entire word.

I would do
`Session *session in sessions`

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 10]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
-            NSComparisonResult result = [now compare:sesh.occursAt];
-            
-            if ([currUser.username isEqualToString:user.username] && result == NSOrderedAscending) {
-                [self.sessionList addObject:sesh];
-                break;
-            }
-        }
-    }
-}
+	NSMutableArray *tempList = (NSMutableArray *)sessions;
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:85]: this is not really making a copy. 
You have the same object.  do you want to use [sessions copy]? 
or [NSMutableArray arrayWithArray:sessions]?

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 11]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
-    paragraph.lineBreakMode = NSLineBreakByWordWrapping;
-    paragraph.alignment = NSTextAlignmentCenter;
-    
-    NSDictionary *attributes = @{NSFontAttributeName: [UIFont systemFontOfSize:14.0f],
-                                 NSForegroundColorAttributeName: [UIColor lightGrayColor],
-                                 NSParagraphStyleAttributeName: paragraph};
-                                 
-    return [[NSAttributedString alloc] initWithString:text attributes:attributes];
-}
+			NSDate *now = [NSDate date];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:93]: should now live outside the loop? 
So you can reutilize it?

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 12]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
+			NSComparisonResult result = [now compare:sesh.occursAt];
 
-- (UIColor *)backgroundColorForEmptyDataSet:(UIScrollView *)scrollView
-{
-    return [Constants playmateBlue];
+			if ([currUser.username isEqualToString:user.username] && result == NSOrderedAscending) {
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:96]: to search an object you can also use: `https://developer.apple.com/documentation/foundation/nsarray/1408043-indexofobjectpassingtest`

File: Playmate/View Controllers/Home Tab/HomeViewController.m [hunk 13]
```diff
@@ -26,187 +26,141 @@ @interface HomeViewController () <UITableViewDelegate, UITableViewDataSource, DZ
...
-     }
-     
-     if ([segue.identifier isEqualToString:@"toCalendar"]) {
-         CalendarViewController *VC = [segue destinationViewController];
-         VC.rawSessionList = self.sessionList;
-     }
-     
- }
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+	if ([sender isMemberOfClass:[SessionCell class]]) {
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/HomeViewController.m:144]: should we use the segue identifier?

File: Playmate/View Controllers/Home Tab/MapPinsViewController.m [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    
+    //Create location manager
+    pinLocationManager = [[CLLocationManager alloc] init];
+    [pinLocationManager setDelegate:self];
+    [pinLocationManager requestAlwaysAuthorization];
+    [pinLocationManager startUpdatingLocation];
+    [pinLocationManager startUpdatingHeading];
+    pinLocationManager.distanceFilter = kCLDistanceFilterNone;
+    pinLocationManager.desiredAccuracy = kCLLocationAccuracyBest;
+
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Home Tab/MapPinsViewController.m:45]: Maybe all this should be inside a new function

`initPinLocationManager`

File: Playmate/View Controllers/Profile Tab/Friends/FriendRequestsViewController.m [hunk 15]
```diff
@@ -27,13 +27,9 @@ - (IBAction)switchViewController:(id)sender {
...
-    if (switcher.selectedSegmentIndex == 0) {
-        self.incomingView.alpha = 1;
-        self.outgoingView.alpha = 0;
-    } else {
-        self.incomingView.alpha = 0;
-        self.outgoingView.alpha = 1;
-    }
+    const BOOL isSegmentZeroIndex = (switcher.selectedSegmentIndex == 0);
+    self.incomingView.alpha = isSegmentZeroIndex ? 1 : 0;
+    self.outgoingView.alpha = isSegmentZeroIndex ? 0 : 1;
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Profile Tab/Friends/FriendRequestsViewController.m:32]: 💙

File: Playmate/View Controllers/Profile Tab/Friends/FriendsListViewController.m [hunk 16]
```diff
@@ -33,13 +34,16 @@ - (void)viewDidLoad {
...
-    self.friendsList = thisPc[@"friendsList"];
+    PlayerConnection *playerConnection = [Constants getPlayerConnectionForUser:self.thisUser];
+    
+    self.friendsList = playerConnection[@"friendsList"];
     
     // set up refresh control
     self.refreshControl = [[UIRefreshControl alloc] init];
-    [self.refreshControl addTarget:self action:@selector(fetchData) forControlEvents:UIControlEventValueChanged];
+    [self.refreshControl addTarget:self
+                         action:@selector(fetchData)
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Profile Tab/Friends/FriendsListViewController.m:45]: nice

File: Playmate/View Controllers/Profile Tab/Friends/IncomingRequestsViewController.m [hunk 17]
```diff
@@ -37,7 +38,9 @@ - (void)viewWillAppear:(BOOL)animated {
     
     // set up refresh control
     self.refreshControl = [[UIRefreshControl alloc] init];
-    [self.refreshControl addTarget:self action:@selector(fetchData) forControlEvents:UIControlEventValueChanged];
+    [self.refreshControl addTarget:self
+                         action:@selector(fetchData)
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Profile Tab/Friends/IncomingRequestsViewController.m:42]: usually, when you break a function to several lines, you want to keep the `:` aligned in the same column. 

I am happy to jump into a VC, to explain myself better.

File: Playmate/View Controllers/Profile Tab/Friends/OutgoingRequestsViewController.m [hunk 18]
```diff
@@ -75,24 +74,17 @@ - (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
 
 #pragma mark - Empty table view protocol methods
 
-- (UIImage *)imageForEmptyDataSet:(UIScrollView *)scrollView
-{
-    return [Constants resizeImage:[UIImage imageNamed:@"empty_friend_request"] withDimension:80];
+- (UIImage *)imageForEmptyDataSet:(UIScrollView *)scrollView {
+    return [UIImage imageNamed:@"logo_small"];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Profile Tab/Friends/OutgoingRequestsViewController.m:78]: We do not need to resize anymore?

File: Playmate/View Controllers/Search Tab/SearchViewController.m [hunk 19]
```diff
@@ -99,38 +100,21 @@ - (void)fetchData {
...
 
-- (NSAttributedString *)descriptionForEmptyDataSet:(UIScrollView *)scrollView
-{
+- (NSAttributedString *)descriptionForEmptyDataSet:(UIScrollView *)scrollView {
     NSString *text = [Constants emptySearchPlaceholderMsg];
-    
-    NSMutableParagraphStyle *paragraph = [NSMutableParagraphStyle new];
-    paragraph.lineBreakMode = NSLineBreakByWordWrapping;
-    paragraph.alignment = NSTextAlignmentCenter;
-
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Search Tab/SearchViewController.m:124]: nice!

File: Playmate/View Controllers/Search Tab/SearchViewController.m [hunk 20]
```diff
@@ -243,7 +227,11 @@ - (void)fetchDataWithFilters:(Filters *)filter {
     [query findObjectsInBackgroundWithBlock:^(NSArray *sessions, NSError *error) {
         if (sessions != nil) {
             
-            self.sessionList = [self filterSessions:sessions withLocation:filter.location andRadius:filter.radius];
+            if (filter.location != nil) {
+                self.sessionList = [self filterSessions:sessions withLocation:filter.location andRadius:filter.radius];
+            } else {
+                self.sessionList = sessions;
+            }
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Search Tab/SearchViewController.m:234]:
```
self.sessionList =  filter.location != nil ? [self filterSessions:sessions 
                                                                            withLocation:filter.location 
                                                                                andRadius:filter.radius] : nil;
```

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m [hunk 21]
```diff
@@ -124,54 +124,50 @@ - (IBAction)didTapFriend:(id)sender {
...
         
-        // remove me.objectId from self.user.pc friends list
-        PlayerConnection *theirPc = [self.user[@"playerConnection"][0] fetchIfNeeded];
+        // remove user.objectId from self.user.playerconnect friends list
+        PlayerConnection *theirPlayerConnection = [self.user[@"playerConnection"][0] fetchIfNeeded];
         
-        tempFriendsList = (NSMutableArray *)theirPc[@"friendsList"];
-        [tempFriendsList removeObject:me.objectId];
-        theirPc[@"friendsList"] = (NSArray *)tempFriendsList;
+        tempFriendsList = (NSMutableArray *)theirPlayerConnection[@"friendsList"];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:139]: Love the renaming. 

Still one question. 

I see in line 128 and in line 137
`PlayerConnection *myPlayerConnection = [user[@"playerConnection"][0] fetchIfNeeded];`
`PlayerConnection *theirPlayerConnection = [self.user[@"playerConnection"][0] fetchIfNeeded];`

what is the difference between `myPlayerConnection` and `theirPlayerConnection`

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m [hunk 22]
```diff
@@ -124,54 +124,50 @@ - (IBAction)didTapFriend:(id)sender {
     
     if (self.isMyFriend) {
         // if removing friend
-        
-        PFUser *me = [[PFUser currentUser] fetchIfNeeded];
-        
         // remove self.user.objectId from my friends list
-        PlayerConnection *myPc = [me[@"playerConnection"][0] fetchIfNeeded];
+        PlayerConnection *myPlayerConnection = [user[@"playerConnection"][0] fetchIfNeeded];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:128]: I think you could use: `getPlayerConnectionForUser`

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m [hunk 23]
```diff
@@ -124,54 +124,50 @@ - (IBAction)didTapFriend:(id)sender {
...
+            playerConnection = [PlayerConnection initializePlayerConnection];
             
-            NSMutableArray *tempPendingList = (NSMutableArray *)pc.pendingList;
+            NSMutableArray *tempPendingList = (NSMutableArray *)playerConnection.pendingList;
             [tempPendingList addObject:self.user.objectId];
-            pc.pendingList = (NSArray *)tempPendingList;
+            playerConnection.pendingList = (NSArray *)tempPendingList;
         } else {
-            pc = [user[@"playerConnection"][0] fetchIfNeeded];
+            playerConnection = [user[@"playerConnection"][0] fetchIfNeeded];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:160]: could you use here? `getPlayerConnectionForUser` ?

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 24]
```diff
@@ -33,11 +34,14 @@ @interface SessionDetailsViewController () <UICollectionViewDelegate, UICollecti
 @implementation SessionDetailsViewController
 
 PFUser *me;
+BOOL partOfSession;
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:37]: devs name booleans as if the were asking questions for example here:
`isPartOfSession`

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 25]
```diff
@@ -161,27 +145,94 @@ - (void)stopConfetti {
...
+                    userToRemove = user;
+                    break;
+                }
+            }
+            
+            [oldPlayersList removeObject:userToRemove];
+                        
+            session[@"playersList"] = (NSArray *)oldPlayersList;
+            
+            int newOccupied = (int)oldPlayersList.count;
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:173]: do you need to cast to int?

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 26]
```diff
@@ -161,27 +145,94 @@ - (void)stopConfetti {
...
+        PFQuery *query = [PFQuery queryWithClassName:@"SportsSession"];
+
+        // Retrieve the object by id
+        [query getObjectInBackgroundWithId:self.sessionDetails.objectId
+                                     block:^(PFObject *session, NSError *error) {
+            // Add myself to the session
+            NSMutableArray *oldPlayersList = (NSMutableArray *)session[@"playersList"];
+            
+            PFUser * _Nullable userToRemove = nil;
+            for (PFUser *user in oldPlayersList) {
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:162]: https://developer.apple.com/documentation/foundation/nsarray/1408043-indexofobjectpassingtest

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 27]
```diff
@@ -161,27 +145,94 @@ - (void)stopConfetti {
...
+            // Add myself to the session
+            NSMutableArray *oldPlayersList = (NSMutableArray *)session[@"playersList"];
+            [oldPlayersList addObject:me];
+            
+            session[@"playersList"] = (NSArray *)oldPlayersList;
+            
+            int newOccupied = (int)oldPlayersList.count;
+            session[@"occupied"] = [NSNumber numberWithInt:newOccupied];
+            
+            [session saveInBackground];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:199]: seems like this 3 lines are also repeated in line 173

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 28]
```diff
@@ -161,27 +145,94 @@ - (void)stopConfetti {
...
+        }];
+    }
+}
+
+- (void)updateJoinUi {
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:204]: usually as UI is an acronym, we keep it in capital letters

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m [hunk 29]
```diff
@@ -191,10 +242,9 @@ - (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSe
 }
 
 - (nonnull __kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
-    
     PlayerProfileCollectionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"PlayerProfileCollectionCell" forIndexPath:indexPath];
     
-    cell.userProfile = self.sessionDetails.playersList[indexPath.row];
+    cell.userProfile = self.sessionDetails.playersList[self.sessionDetails.playersList.count- indexPath.row-1];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:247]: why the change here? 

Any reason for the -1.
Maybe add a comment, so you will remember why.

Are we confident we will not have an index out of bounds? or a negative number?

File: Playmate/Utilities/Constants.m [hunk 30]
```diff
@@ -23,15 +23,24 @@ + (NSString *)decathalonSportsListString {
     return @"https://sports.api.decathlon.com/sports";
 }
 
++ (NSString *)decathalonOneSportString {
+    return @"https://sports.api.decathlon.com/sports/:";
+}
+
+// For Parse
++ (PlayerConnection *)getPlayerConnectionForUser:(PFUser *)user {
+    return [user[@"playerConnection"][0] fetchIfNeeded];
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/Utilities/Constants.m:32]: Hmm I'm not quite sure if I can add a method in user since its a pre defined Parse PFUser object. I'll think of a better way, but for now i'll move it to Helpers file instead of constants

File: Playmate/Utilities/APIManager.m [hunk 31]
```diff
@@ -107,4 +117,29 @@ - (void)getGeocodedLocation:(NSString *)address withCompletion:(void(^)(Location
...
+    NSURLRequest *request = [NSURLRequest requestWithURL:url cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:10.0];
+    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+        if (error != nil) {
+            NSLog(@"%@", [error localizedDescription]);
+            completion(nil, error);
+        }
+        else {
+            NSDictionary *dataDictionary = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
+            // parse dictionary
+            NSDictionary *result = dataDictionary[@"features"][0][@"properties"];
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/Utilities/APIManager.m:137]: as long as a latitude and longitude are sent through the reverse geocoding dictionary, a non empty result dictionary is returned with a features array containing dictionaries. Does this situation still call for needing error handling?

File: Playmate/Utilities/Constants.m [hunk 32]
```diff
@@ -23,15 +23,24 @@ + (NSString *)decathalonSportsListString {
...
 
 + (NSDate *) dateWithHour:(NSInteger)hour
                   minute:(NSInteger)minute
                   second:(NSInteger)second
                  fromDate:(NSDate *)date
 {
-   NSCalendar *calendar = [NSCalendar currentCalendar];
-   NSDateComponents *components = [calendar components: NSCalendarUnitYear|
+    NSCalendar *calendar = [NSCalendar currentCalendar];
+    NSDateComponents *components = [calendar components: NSCalendarUnitYear|
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/Utilities/Constants.m:43]: Nope, i'm unsure why this happens sometimes... i'll be on the lookout

File: Playmate/View Controllers/Home Tab/HomeViewController.m (hunk 7 above)
Comment (jociet101) [reply Playmate/View Controllers/Home Tab/HomeViewController.m:17]: Oops, this was a deletion in progress, the table view was not working at the time and I have since completely chaged the layout and implemented a collection view. no worries on this

File: Playmate/View Controllers/Home Tab/HomeViewController.m (hunk 8 above)
Comment (jociet101) [reply Playmate/View Controllers/Home Tab/HomeViewController.m:72]: that's weird. Looks fine to me on xcode

File: Playmate/View Controllers/Profile Tab/Friends/IncomingRequestsViewController.m (hunk 17 above)
Comment (jociet101) [reply Playmate/View Controllers/Profile Tab/Friends/IncomingRequestsViewController.m:42]: Ahh that makes a lot of sense. I understand what you mean by aligning the colons, but was there anything else you wanted to explain more in depth?

File: Playmate/View Controllers/Session Details View/SessionDetailsViewController.m (hunk 25 above)
Comment (jociet101) [reply Playmate/View Controllers/Session Details View/SessionDetailsViewController.m:173]: array.count returns a long so I use this to get rid of warning. Hopefully there is not over 2147483647 people participating in one session :)

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m (hunk 21 above)
Comment (jociet101) [reply Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:139]: Ah, my PlayerConnection model is used to represent my friend network and every PFUser(player) has their own PlayerConnection object connected to the PFUser's objectId. My friends network an undirected graph and to store data in parse, I use an adjacency list representation such that if Player A and Player B are friends, then B is in A's friend list and A is in B's friend list. That is why there is myPlayerConnection (let's say I am player A, then I need my PC to add B to the friend list) and theirPlayerConnection (to add me, player A, to Player B's friend list).
